src/wasm-lib/kcl/src/execution/memory.rs

```diff
@@ -1,74 +1,405 @@
+        // waste time and space on a new env. Using usize::MAX means we'll get an overflow if we
+        // try to access anything rather than a silent error.
+        self.current_env = EnvironmentRef(usize::MAX, SnapshotRef::none());
+    }
+
+    /// Push a new stack frame on to the call stack with no connection to a parent environment.
+    ///
+    /// Suitable for executing a separate module.
+    pub fn push_new_root_env(&mut self) {
+        self.stats.env_count += 1;
+
+        self.call_stack.push(self.current_env);
+        let new_env = Environment::new_root();
+        self.current_env = EnvironmentRef(self.environments.len(), SnapshotRef::none());
+        self.environments.push(new_env);
+    }
+
+    /// Pop a frame from the call stack and return a reference to the popped environment. The popped
+    /// environment is preserved if it may be referenced (so the returned reference will remain valid).
+    ///
+    /// The popped environment may be retained completely (if it may be referenced by a function decl
+    /// or import) or retained but its contents deleted or completely discarded.
+    pub fn pop_env(&mut self) -> EnvironmentRef {
+        let old = self.current_env;
+        self.current_env = self.call_stack.pop().unwrap();
+
+        if !old.is_rust_env() {
+            self.environments[old.index()].compact();
+
+            if self.environments[old.index()].is_empty() {
+                if old.index() == self.environments.len() - 1 {
+                    self.environments.pop();
+                    self.stats.env_gcs += 1;
+                } else {
+                    self.stats.skipped_env_gcs += 1;
+                }
+            } else {
+                self.stats.preserved_envs += 1;
+            }
+        }
+
+        old
+    }
+
+    /// Pop a frame from the call stack and return a reference to the popped environment. The popped
+    /// environment is always preserved.
+    pub fn pop_and_preserve_env(&mut self) -> EnvironmentRef {
+        let old = self.current_env;
+        self.current_env = self.call_stack.pop().unwrap();
+        old
+    }
+
+    /// Merges the specified environment with the current environment, rewriting any environment refs
+    /// taking snapshots into account. Deletes (if possible) or clears the squashed environment.
+    pub fn squash_env(&mut self, old: EnvironmentRef) {
+        assert!(!old.is_rust_env());
+
+        let mut old_env = if old.index() == self.environments.len() - 1 {
+            // Common case and efficient
+            self.stats.env_gcs += 1;
+            self.environments.pop().unwrap()
+        } else {
+            // Should basically never happen in normal usage.
+            self.stats.skipped_env_gcs += 1;
+            let cloned = self.environments[old.index()].clone();
+            self.environments[old.index()] = Environment::new(self.current_env);
+            cloned
```

Comment:
Should be able to avoid the clone with this.

```suggestion
            std::mem::replace(&mut self.environments[old.index()], Environment::new(self.current_env))
```